### DISCUSSION TITLE
R - TEAL on Local Machine (Error)

### DISCUSSION THREAD

**Discussion by BhargavAchanta** [insightsengineering/teal#1086] [General]
I am using the following code to test on Rstudio on my local machine, but the get the following error

 library(teal)
library(shiny)
app <- init(
          data = teal_data(
           dataset("iris",iris),
           dataset("mtcars",mtcars)
          )
          ,
            modules = list(
              module(
                label = "My",
                ui = function(id){
                  tags$div("Hello")
                }
              )
            )
)
            

shinyApp(app$ui,app$server)

Error:
! `to_relational_data()` was deprecated in teal.data 0.4.0 and is now defunct.
ℹ Find more information on https://github.com/insightsengineering/teal/discussions/945
---
Backtrace:
    ▆
 1. └─teal::init(...)
 2.   └─teal.data::to_relational_data(data = data)
 3.     └─teal.data:::.deprecate_function("to_relational_data()", deprecation_detail)
 4.       └─lifecycle::deprecate_stop(when = "0.4.0", what = what, details = details)
 5.         └─lifecycle:::deprecate_stop0(msg)
Run rlang::last_trace(drop = FALSE) to see 1 hidden frame.

**Comment by donyunardi** [comment]
Hi @BhargavAchanta 

Can I see sessionInfo()? 
We're in the middle of releasing teal frameworks, some are already released but some are still in process.
Because of this, you may need to install some packages (like teal.slice and teal) from `main` instead of latest release.

If it's easier, you can wait until next week when we're done releasing everything.

**Reply by BhargavAchanta** [reply]
Here is the session info details. Thank you!

![Capture](https://github.com/insightsengineering/teal/assets/158379030/0ec2e655-57e2-47b7-9163-7babf7cd2c0b)

**Reply by donyunardi** [reply]
As I suspected, you are using the latest release of `teal` and `teal.slice`, you will need to install from `main` branch for these two packages.

```r
install.packages("teal.slice", repos = c("https://insightsengineering.r-universe.dev", getOption("repos")))
install.packages("teal", repos = c("https://insightsengineering.r-universe.dev", getOption("repos")))
```

And, you have to use the new way to define teal.data as @chlebowa has pointed out on different comment.

**Reply by BhargavAchanta** [reply]
Hi Donny, 

i have tried this as well and have the same error 

```
install.packages("teal.slice", repos = c("https://insightsengineering.r-universe.dev", getOption("repos")))
install.packages("teal", repos = c("https://insightsengineering.r-universe.dev", getOption("repos")))


library(teal)
library(shiny)

app <- init(
  data = teal_data(
    "iris" = iris,
    "mtcars" = mtcars
  ),
  modules = list(
    module(
      label = "My",
      ui = function(id) {
        tags$div("Hello")
      }
    )
  )
)

shinyApp(app$ui,app$server)




R version 4.3.2 (2023-10-31 ucrt) -- "Eye Holes"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> install.packages("teal.slice", repos = c("https://insightsengineering.r-universe.dev", getOption("repos")))
WARNING: Rtools is required to build R packages but is not currently installed. Please download and install the appropriate version of Rtools before proceeding:

https://cran.rstudio.com/bin/windows/Rtools/
Installing package into ‘C:/Users/BhargavAchanta/AppData/Local/R/win-library/4.3’
(as ‘lib’ is unspecified)

  There is a binary version available but the source version is later:
           binary source needs_compilation
teal.slice  0.4.0  0.5.0             FALSE

installing the source package ‘teal.slice’

trying URL 'https://cran.rstudio.com/src/contrib/teal.slice_0.5.0.tar.gz'
Content type 'application/x-gzip' length 1314592 bytes (1.3 MB)
downloaded 1.3 MB

* installing *source* package 'teal.slice' ...
** package 'teal.slice' successfully unpacked and MD5 sums checked
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (teal.slice)

The downloaded source packages are in
	‘C:\Users\BhargavAchanta\AppData\Local\Temp\RtmpCYEn1D\downloaded_packages’
> install.packages("teal", repos = c("https://insightsengineering.r-universe.dev", getOption("repos")))
WARNING: Rtools is required to build R packages but is not currently installed. Please download and install the appropriate version of Rtools before proceeding:

https://cran.rstudio.com/bin/windows/Rtools/
Installing package into ‘C:/Users/BhargavAchanta/AppData/Local/R/win-library/4.3’
(as ‘lib’ is unspecified)
trying URL 'https://insightsengineering.r-universe.dev/bin/windows/contrib/4.3/teal_0.14.0.zip'
Content type 'application/zip' length 1312999 bytes (1.3 MB)
downloaded 1.3 MB

package ‘teal’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\BhargavAchanta\AppData\Local\Temp\RtmpCYEn1D\downloaded_packages
> 
> 
> library(teal)
Loading required package: shiny
Loading required package: teal.data
Loading required package: teal.code
Loading required package: teal.slice
Loading required package: teal.transform
Loading required package: magrittr

You are using teal version 0.14.0

Attaching package: ‘teal’

The following objects are masked from ‘package:teal.slice’:

    as.teal_slices, teal_slices

The following object is masked from ‘package:teal.data’:

    get_join_keys

> library(shiny)
> 
> app <- init(
+   data = teal_data(
+     "iris" = iris,
+     "mtcars" = mtcars
+   ),
+   modules = list(
+     module(
+       label = "My",
+       ui = function(id) {
+         tags$div("Hello")
+       }
+     )
+   )
+ )
Error:
! `to_relational_data()` was deprecated in teal.data 0.4.0 and is now defunct.
ℹ Find more information on https://github.com/insightsengineering/teal/discussions/945
Run `rlang::last_trace()` to see where the error occurred.
> rlang::last_trace()
<error/lifecycle_error_deprecated>
Error:
! `to_relational_data()` was deprecated in teal.data 0.4.0 and is now defunct.
ℹ Find more information on https://github.com/insightsengineering/teal/discussions/945
---
Backtrace:
    ▆
 1. └─teal::init(...)
 2.   └─teal.data::to_relational_data(data = data)
 3.     └─teal.data:::.deprecate_function("to_relational_data()", deprecation_detail)
 4.       └─lifecycle::deprecate_stop(when = "0.4.0", what = what, details = details)
 5.         └─lifecycle:::deprecate_stop0(msg)
Run rlang::last_trace(drop = FALSE) to see 1 hidden frame.
```

**Reply by donyunardi** [reply]
Apologies, I gave the wrong URL to get the `main` branch of teal.
Your `teal.slice` version is correct.

Can you please reinstall `teal` again using this code:
```
install.packages("teal", repos = c("https://pharmaverse.r-universe.dev", getOption("repos")))
```

and try it out again?

**Reply by BhargavAchanta** [reply]
Perfect. Works now. Thank you so much to you and to your team for your awesome work and presentations on TEAL package.

**Comment by chlebowa** [comment]
Hi @BhargavAchanta 

Try this:
```
app <- init(
  data = teal_data(
    "iris" = iris,
    "mtcars" = mtcars
  ),
  modules = list(
    module(
      label = "My",
      ui = function(id) {
        tags$div("Hello")
      }
    )
  )
)
```

**Reply by BhargavAchanta** [reply]
Hello, I have tried the code you provided , I have the same error.

Thank you

**Comment by BhargavAchanta** [comment]
Hello,
using the following code to reproduce in locol machine but i see this
error.

 [INFO] 2024-02-07 13:09:26.8288 pid:22628 token:[] teal.modules.clinical
Initializing tm_t_events

Error in choices_selected(c("ARM", "ARMCD"), "ARM") :
  could not find function "choices_selected"




adsl <- tmc_ex_adsl
adae <- tmc_ex_adae

app <- teal::init(
  data = cdisc_data("ADSL" =  adsl,
                   "ADAE" =  adae
  ),
  modules = modules(
    tm_t_events(
      label = "Adverse Event Table",
      dataname = "ADAE",
      arm_var = choices_selected(c("ARM", "ARMCD"), "ARM"),
      llt = choices_selected(
        choices = variable_choices(adae, c("AETERM", "AEDECOD")),
        selected = c("AEDECOD")
      ),
      hlt = choices_selected(
        choices = variable_choices(adae, c("AEBODSYS", "AESOC")),
        selected = "AEBODSYS"
      ),
      add_total = TRUE,
      event_type = "adverse event"
    )
  )
)

if (interactive()) {
  shinyApp(app$ui, app$server)
}

**Reply by donyunardi** [reply]
Can you make sure that you have teal.transform and tmc from `main`?

```r
install.packages("teal", repos = c("https://pharmaverse.r-universe.dev", getOption("repos")))
install.packages("teal.modules.clinical", repos = c("https://pharmaverse.r-universe.dev", getOption("repos")))
```

Again, apologies for this, we're in the middle of the release.

**Comment by BhargavAchanta** [comment]
I will wait until the release. I have been having issues with running all
the example code on the local machine.. sometimes the shiny app appears and
closes immediately and some errors due to functions being deprecated . They
do run fine without any issues on the TEAL playground on posit cloud. I am
installing from main.

Thanks

**Reply by averissimo** [reply]
Apologies that you are experiencing this. Although we are keeping a migration path for the old data APIs, the whole {teal} environment needs to use the CRAN releases _(or development/`main` for those that are not yet on CRAN)_.

As @donyunardi suggested, installing packages from the `pharmaverse.r-universe.dev` repository should allow you to run that `{teal}` application and other examples without issues.

I use the following snippet to make sure that my whole teal environment is up-to-date with development packages.

```r
c(
  "teal", "teal.data", "teal.code", "teal.slice",
  "teal.logger", "teal.widgets", "teal.reporter",
  "teal.transform", "teal.modules.general", "teal.modules.clinical",
  "rtables", "tern"
) |> install.packages(
  repos = c(
    pharmaverse = "https://pharmaverse.r-universe.dev",
    getOption("repos")
  )
)
```

With this and the code snippet you shared above, I can run locally the `{teal}` application _(added just 1 library call)_

All the examples imply that you are loading `{teal.modules.clinical}` so it is fundamental as it will, in turn, load the `choices_selected` and `variable_choices` from `{teal.transform}`.

```r
library(teal.modules.clinical)

adsl <- tmc_ex_adsl
adae <- tmc_ex_adae

app <- teal::init(
  data = cdisc_data("ADSL" =  adsl,
                    "ADAE" =  adae
  ),
  modules = modules(
    tm_t_events(
      label = "Adverse Event Table",
      dataname = "ADAE",
      arm_var = choices_selected(c("ARM", "ARMCD"), "ARM"),
      llt = choices_selected(
        choices = variable_choices(adae, c("AETERM", "AEDECOD")),
        selected = c("AEDECOD")
      ),
      hlt = choices_selected(
        choices = variable_choices(adae, c("AEBODSYS", "AESOC")),
        selected = "AEBODSYS"
      ),
      add_total = TRUE,
      event_type = "adverse event"
    )
  )
)

if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

![image](https://github.com/insightsengineering/teal/assets/211358/63130662-f32b-435b-9942-c222bf4a1591)

**Comment by BhargavAchanta** [comment]
Thank you. runs fine without errors, but i do see the following and the
shinyapp flashes for a sec and closes immediately., not sure why..

Listening on http://127.0.0.1:3038[INFO] 2024-02-08 07:44:58.3513
pid:27528 token:[90831486] teal Initializing
reporter_previewer_moduleWarning: Error in add_card_button_srv:
Assertion on 'names(formals(card_fun))' failed: Must be a subset of
{'card','comment'}, but has additional elements {'label'}.  3: runApp
  2: print.shiny.appobj
  1: <Anonymous>
